src/module.c

```diff
@@ -7418,6 +7444,215 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+ *                   RedisModule_FreeModuleUser(user);
+ *               }
+ *               const char *err_msg = "Authentication denied by Module.";
+ *               *err = RedisModule_CreateString(ctx, err_msg, strlen(err_msg));
+ *               return REDISMODULE_AUTH_HANDLED;
+ *           }
+ *           return REDISMODULE_AUTH_NOT_HANDLED;
+ *       }
+ *
+ *       int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+ *           REDISMODULE_NOT_USED(argv);
+ *           REDISMODULE_NOT_USED(argc);
+ *           if (RedisModule_Init(ctx,"authmodule",1,REDISMODULE_APIVER_1)== REDISMODULE_ERR)
+ *               return REDISMODULE_ERR;
+ *           RedisModule_RegisterCustomAuthCallback(ctx, auth_cb);
+ *           return REDISMODULE_OK;
+ *       }
+ */
+void RM_RegisterCustomAuthCallback(RedisModuleCtx *ctx, RedisModuleCustomAuthCallback cb) {
+    RedisModuleCustomAuthCtx *auth_ctx = zmalloc(sizeof(RedisModuleCustomAuthCtx));
+    auth_ctx->module = ctx->module;
+    auth_ctx->auth_cb = cb;
+    listAddNodeTail(moduleCustomAuthCallbacks, auth_ctx);
+}
+
+/* Helper function to invoke the free private data callback of a Module blocked client. */
+void moduleInvokeFreePrivDataCallback(client *c, RedisModuleBlockedClient *bc) {
+    if (bc->privdata && bc->free_privdata) {
+        RedisModuleCtx ctx;
+        int ctx_flags = c == NULL ? REDISMODULE_CTX_BLOCKED_DISCONNECTED : REDISMODULE_CTX_NONE;
+        moduleCreateContext(&ctx, bc->module, ctx_flags);
+        ctx.blocked_privdata = bc->privdata;
+        ctx.client = bc->client;
+        bc->free_privdata(&ctx,bc->privdata);
+        moduleFreeContext(&ctx);
+    }
+}
+
+/* Returns 1 if the module has any custom auth callbacks registered. If `should_unregister` is
+ * non-zero, all the module's custom auth cbs are unregistered.
+ * Returns 0 otherwise - if the Module has no custom auth cb registered. */
+int moduleSearchOrUnregisterCustomAuthCBs(RedisModule *module, int should_unregister) {
+    int callback_found = 0;
+    listIter li;
+    listNode *ln;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    while ((ln = listNext(&li))) {
+        RedisModuleCustomAuthCtx *ctx = listNodeValue(ln);
+        if (ctx->module == module) {
+            callback_found = 1;
+            if (!should_unregister) return callback_found;
+            listDelNode(moduleCustomAuthCallbacks, ln);
+            zfree(ctx);
+        }
+    }
+    return callback_found;
+}
+
+/* Returns 1 if custom module based auth is in progress on any client. Returns 0 otherwise. */
+int isCustomAuthInProgress() {
+    if (!listLength(moduleCustomAuthCallbacks)) return 0;
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients,&li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *c = listNodeValue(ln);
+        if (clientHasModuleAuthInProgress(c)) {
+            return 1;
+        }
+    }
```

Comment:
that look is capable of being huge (possibly 30k clients).
if we need that capability, we'll have to add a list of clients that are mid auth.

Reply:
With this (https://github.com/redis/redis/pull/11659#discussion_r1125084544), this loop will be removed

Reply:
This is removed in this commit: https://github.com/redis/redis/pull/11659/commits/9f073e1bee09307ae03cac564ee11e5742a3472d